src/banded/BandedMatrix.jl

```diff
@@ -65,7 +65,11 @@ BandedMatrix(M::BandedMatrix{V}) where {V} =
 convert(::Type{BandedMatrix{V}}, M::BandedMatrix{V}) where {V} = M
 convert(::Type{BandedMatrix{V}}, M::BandedMatrix) where {V} =
         _BandedMatrix(convert(AbstractMatrix{V}, M.data), M.m, M.l, M.u)
-convert(::Type{BandedMatrix}, M::BandedMatrix{V}) where {V} = M
+convert(::Type{BandedMatrix}, M::BandedMatrix) = M
+function convert(BM::Type{<: BandedMatrix}, M::BandedMatrix)
+    M.data isa fieldtype(BM, :data) && return M
+    _BandedMatrix(convert(fieldtype(BM, :data), M.data), M.m, M.l, M.u)
+end
```

Comment:
This needs to be rewritten: it's not type stable and won't necessarily convert to the right type. I think it's better to just add 
```julia
convert(BM::Type{BandedMatrix{T,V}}, M::BandedMatrix) where V<:AbstractMatrix{T} where T =
          _BandedMatrix(convert(AbstractMatrix{V}, M.data), M.m, M.l, M.u)
```

Reply:
Actually, I think this is type stable (at least in v0.6). You will find the relevant tests are sprinkled with `@inferred` specifically to check this.  The likely reason is that the branching involves compile-time (and specialization-time) constants.

The main reason for the separate returns is that some types return an object different from the original one even when unnecessary (specifically, JLArray does).
Finally, this function is only called when an explicit specialization over the container type is used. Hence your example is inadequate since it would not convert the type of container.

To be more explicit, the folllowing conversions are type stable:
  - BandedMatrix{..., Matrix{..}} -> BandedMatrix{..., Matrix} 
  - BandedMatrix{..., JLArray{..., 2}} -> BandedMatrix{..., JLArray} (see test_banded.jl:276)
  - BandedMatrix{..., JLArray{..., 2}} -> BandedMatrix{..., Matrix} (see test_banded.jl:281)
  - BandedMatrix{Int32, Matrix{Int32}} -> BandedMatrix{Int64, JLArray{Int64, 2}}

Above, (something  -> BandedMatrix{..., Y}) type stability is preserved whether Y is concrete or abstract.

Whereas the following is not:
  - BandedMatrix{..., Matrix} -> BandedMatrix{..., JLArrray}
  - BandedMatrix{..., Matrix} -> BandedMatrix{..., JLArrray{T, 2}}

But that's because the conversion (concrete Matrix) -> (abstract JLArray) is not type stable either. Not much we can do there.

So, overall I think the current version is still better than the one in your comment. Let me know if you still want the change implemented.

Reply:
OK, if the tests pass then I guess it's fine.